### ISSUE TITLE
db/test/db_metadata uses a mix of err == nil and err == ErrNotFound when testing for not found

### ISSUE THREAD

**Issue by andresrinivasan** [edgexfoundry/edgex-go#1203]
I worked around this in db/redis for now but db_metatdata.go needs a cleanup pass which will cascade into both db/redis and db/mongo changes.

**Comment by jduranf** [comment]
It's not clear for me if this is an issue or not...

If you call `GetProvisionWatchersByProfileId(id string)`, if the profile doesn't exit you should receive an `err == ErrNotFound`, but if the profile exists but there isn't any provision watcher using this profile, you should receive `err == nil`.

In other calls like `GetProvisionWatchersByIdentifier(k string, v string)`, if there isn't any watcher with the requested identifier you should receive `err == nil`.

This is why there is a mix of `err == nil` and `err == ErrNotFound`.

**Comment by TomFlem** [comment]
I have run into this issue when updating C Device Services.

During startup, after registering the device service with metadata, a Device Service makes a call to metadata to check whether any devices exist for its device service name.

Metadata calls, `restGetDeviceByServiceName()` which, if the DS exists, calls `GetDevicesByServiceId()`.


The Mongo implementation is:

```
func (mc MongoClient) GetDevicesByServiceId(id string) ([]contract.Device, error) {
  ds, err := mc.getDeviceServiceById(id)
    if err != nil {
        return []contract.Device{}, err
    }
  return mc.getDevices(bson.M{"service.$id": ds.Id})
}

```
Get the DS, if it doesn't exist error, then query for registered devices and return.


The Redis implementation is:

```
func (c *Client) GetDevicesByServiceId(id string) ([]contract.Device, error) {
  d, err := c.getDevicesByValue(db.Device + ":service:" + id)

    // XXX This is here only because test/db_metadata.go is inconsistent when testing for _not found_. It
    // should always be checking for database.ErrNotFound but too often it is checking for nil
    if len(d) == 0 {
        err = dataBase.ErrNotFound
    }

  return d, err
}
```

Get the devices, if none are returned assume that the DS is not found.

This causes the C SDK to fail on launch, as even if the DS exists, if no devices are registered ErrNotFound is returned.


Shouldn't the Redis implementation also check if the DS exists before querying for devices? See:

```
func (c *Client) GetDevicesByServiceId(id string) ([]contract.Device, error) {
  _, err := c.GetDeviceServiceById(id)
  if err != nil {
        return []contract.Device{}, err
    }
    d, err := c.getDevicesByValue(db.Device + ":service:" + id)
    return d, err
}
```

**Comment by andresrinivasan** [comment]
> It's not clear for me if this is an issue or not...
> 
> If you call `GetProvisionWatchersByProfileId(id string)`, if the profile doesn't exit you should receive an `err == ErrNotFound`, but if the profile exists but there isn't any provision watcher using this profile, you should receive `err == nil`.
> 
> In other calls like `GetProvisionWatchersByIdentifier(k string, v string)`, if there isn't any watcher with the requested identifier you should receive `err == nil`.
> 
> This is why there is a mix of `err == nil` and `err == ErrNotFound`.

@jduranf This is exactly what I find confusing: why do we have two possible results for not found? If we can decide and document what nil means and what ErrNotFound means then we can propagate that to the tests and make the code consistent.

**Comment by SteveOss** [comment]
Think is actually clear, not finding any associated devices is a normal return case. Using  ErrNotFound
should only be for when there is an **error** case. Querying an extent that can be empty is not an error, so
the error status code should not be returned.

**Comment by andresrinivasan** [comment]
> Think is actually clear, not finding any associated devices is a normal return case. Using ErrNotFound
> should only be for when there is an **error** case. Querying an extent that can be empty is not an error, so
> the error status code should not be returned.

@SteveOss given

```go
deviceServices, err = db.GetDeviceServicesByAddressableId(uuid.New().String())
```

Should `err == nil && len(deviceServices) == 0`?

If so, the db tests need to be revised. And of course I'll update the Redis impl.

**Comment by SteveOss** [comment]
@andresrinivasan I think it's an error if the addressable can't be found. But not an error if the
addressable can be found but contains no devices.

**Comment by andresrinivasan** [comment]
@SteveOss @jduranf @TomFlem If the above is the agreed definition, I can correct my code.

Where should this be documented? For example, add Go doc to the interface for each method.

**Comment by TomFlem** [comment]
I agree with @SteveOss's definition

**Comment by SteveOss** [comment]
@andresrinivasan I think it's general practice that if you are looking up an extent of B on A. Then if A can't be found, this is an error, however if the extent of B is empty this is not.

**Comment by andresrinivasan** [comment]
Before I start making code changes, I have documented one of the interfaces with methods I intend to verify/update. Please confirm this is aligned with what has been discussed above.

https://github.com/andresrinivasan/edgex-go/tree/issue-1203-cleanup-errnotfound-use
See https://github.com/edgexfoundry/edgex-go/compare/master...andresrinivasan:issue-1203-cleanup-errnotfound-use?expand=1#diff-965b5860c2a9b246c93d5681a29530b7

**Comment by SteveOss** [comment]
@andresrinivasan This all looks good to me.

**Comment by andresrinivasan** [comment]
In the case of `GetDeviceReportsByAction()` or `GetDevicesWithLabel()`, this should never return ErrNotFound; it should return a slice with len >= 0, correct?

**Comment by jduranf** [comment]
Correct

**Comment by andresrinivasan** [comment]
I'm still struggling with establishing a pattern in my brain. Given

```go
	ValueDescriptorByName(name string) (contract.ValueDescriptor, error)

	ValueDescriptorsByName(names []string) ([]contract.ValueDescriptor, error)

	ValueDescriptorById(id string) (contract.ValueDescriptor, error)

	ValueDescriptorsByUomLabel(uomLabel string) ([]contract.ValueDescriptor, error)

	ValueDescriptorsByLabel(label string) ([]contract.ValueDescriptor, error)

	ValueDescriptorsByType(t string) ([]contract.ValueDescriptor, error)
```

`ValueDescriptorByName(name string) (contract.ValueDescriptor, error)` and `ValueDescriptorsByName(names []string) ([]contract.ValueDescriptor, error)` feels like they should return _ErrNotFound_ but `ValueDescriptorsByUomLabel(uomLabel string) ([]contract.ValueDescriptor, error)` **should not** return _ErrNotFound_. 

Is the pattern then an index (e.g. id, name) based query can result in _ErrNotFound_ and with a label/tag/type query it is acceptable to have 0 results (and no _ErrNotFound_)?

**Comment by jduranf** [comment]
If we are requesting a specific value descriptor, like `ValueDescriptorByName` or `ValueDescriptorById`, then `ErrNotFound` should be returned if not found.

If we are requesting an array of value descriptors, like `ValueDescriptorsByName`, `ValueDescriptorsByUomLabel`, `ValueDescriptorsByLabel` or `ValueDescriptorsByType`, then we will receive an empty array, without `ErrNotFound`.

The exceptions in this second case is when we expect an array, but from a specific item which my not exist, like `GetProvisionWatchersByProfileId`. In this case we return `ErrNotFound` if the profile doesn't exist, or an empty array if the profile exist but there isn't any provision watcher.

**Comment by andresrinivasan** [comment]
Thanks. I appreciate you breaking it down.

**Comment by difince** [comment]
I just came across this problem. Because of it I cannot clear my database when using Redis database. Isn't it good to be fix for the current release (edinburg)?

**Comment by andresrinivasan** [comment]
The larger set of updates are not going to make it in. We can create a specific issue for what you found and correct that. Can you elaborate on a new issue?

**Comment by difince** [comment]
1. DeviceService [cannot be deleted](https://github.com/edgexfoundry/edgex-go/blob/8bfb242945e30f6b8a24071db0a838d0bcc60c0b/internal/core/metadata/rest_deviceservice.go#L370), if No [Devices are assigned ](https://github.com/edgexfoundry/edgex-go/blob/8bfb242945e30f6b8a24071db0a838d0bcc60c0b/internal/pkg/db/redis/metadata.go#L292)to it.

1. DeviceProfiles [cannot be deleted](https://github.com/edgexfoundry/edgex-go/blob/8bfb242945e30f6b8a24071db0a838d0bcc60c0b/internal/core/metadata/rest_deviceprofile.go#L356), if the DeviceProfile have No [ Provisioning watchers](https://github.com/edgexfoundry/edgex-go/blob/8bfb242945e30f6b8a24071db0a838d0bcc60c0b/internal/pkg/db/redis/metadata.go#L1008)

and as a site effect -> Addressables cannot be deleted, because we do not clean the DeviceService (1) and the addressables are still in [use](https://github.com/edgexfoundry/edgex-go/blob/8bfb242945e30f6b8a24071db0a838d0bcc60c0b/internal/core/metadata/rest_addressable.go#L141)  

All these cleaning issues have the root cause described above: when searching for _multiple_ objects  from the DB by some criteria, and no object was found, this is not a condition for raising an error, instead an empty list should be returned. [described here](https://github.com/edgexfoundry/edgex-go/issues/1203#issuecomment-488569660)

**Comment by andresrinivasan** [comment]
Ok. That's bigger than I expected. I'm not going to be able to context switch to this until next week. Do you have any cycles so we can divide computer conquer?

**Comment by difince** [comment]
@andresrinivasan I do not have anything urgent, so I will be glad if we work together

**Comment by tsconn23** [comment]
I want to ask, are we getting any value from this? `db/test/db_metadata`

It sounds like we're spending a lot of cycles on this, keeping it inline as we make other changes. Does this add to our test coverage metrics? Does the inclusion of this in the project give us any additional certainty w/r/t the correct operation of the platform? If not and it's just overhead then I'm inclined to remove it. Thoughts?

**Comment by andresrinivasan** [comment]
If db_metadata.go is not properly unit testing Core Metadata, I would prefer we invest in correcting the unit tests as I depend on them for development.

**Comment by tsconn23** [comment]
Unless anyone has input to the contrary, I'm going to create an issue very shortly to whack this entire `internal/pkg/db/test` folder. It's not clear to me that this separate testing strategy is driving any value.

- It's not wired into any kind of CI build script
- It negatively affects our code coverage statistics because `go test` thinks its a code directory without any tests
`?   	github.com/edgexfoundry/edgex-go/internal/pkg/db/test	[no test files]`
- It adds overhead to every PR that involves some change to the DBClient interfaces. Example: https://github.com/edgexfoundry/edgex-go/pull/1496/files#diff-5e33353fc4ef055b0aafa168de0316a0

**Comment by andresrinivasan** [comment]
I use the go tests for unit/smoke tests. I find the blackbox tests to be more difficult to setup and run. If I hit an issue, mapping from the output of blackbox tests to a line number is definitely not as easy as go tests. If I want to attach a debugger, go tests is also easier.

I can't speak to the additional overhead it adds to PRs, but adding go tests to the CI build and having code coverage ignore the test directory seems like something that can be corrected.

**Comment by jduranf** [comment]
I agree with @andresrinivasan, those tests helps a lot to implement a new DB and to check for any regression. Should be better to add the tests to the CI and ignore the test directory for code coverage.

**Comment by tsconn23** [comment]
> Should be better to add the tests to the CI and ignore the test directory for code coverage.
@jduranf @andresrinivasan Can one of you sign up for this action item please?

**Comment by andresrinivasan** [comment]
@jduranf, I'll divide and conquer with you. Which one do you want?

**Comment by jduranf** [comment]
If it's fine, a simple way to ignore the test directory, is to change this line in Makefile:
_GO111MODULE=on go test -cover ./..._
for this one:
_GO111MODULE=on go test \`go list ./... | grep -v internal/pkg/db/test\`_
If it's agreed, I will create the PR.

Another option is to move the files in internal/pkg/db/test to internal/pkg/db/mongo and internal/pkg/db/redis, adding the suffix _test. But this implies duplicate the test files in each db implementation.

Also, I want to notice, that having directories for testing is a common practice, like [httptest](http://godoc.org/net/http/httptest).

**Comment by brandonforster** [comment]
The more I dig into this, the more I think these tests should be removed completely.

- They are integration tests, like the blackbox tests.
- They provide no additional insight beyond the information the blackbox tests provide.
- The problem of the blackbox tests being obtuse and hard to work with is a noted and recognized problem, [but I am doing my best to try and make it less so.](https://github.com/edgexfoundry/blackbox-testing/pull/310)
- These tests are going to require a messy solution to exclude them from coverage and treat them as "normal" tests. They can't be simply added as `_test` files because they'll get run by `make test`, which will of course not work because nothing is running (nor should anything be.)


**tl;dr:** These tests add no value and will take a nontrivial amount of work to do anything meaningful with. I think resources are better focused on making the blackbox tests more friendly.

**Comment by andresrinivasan** [comment]
If you can address the concerns from earlier in the thread wrt being able to easily unit test / smoke test when working in the persistence code, I'm happy to switch to the blackbox tests to as I go through dev/debug cycles.

**Comment by AnthonyMBonafide** [comment]
I think @brandonforster makes some good points here. Even though I think that we should have something which tests the functionality of the `DBClient` implementation, I think we can maybe do this is in a better way. I like that we have tests that ensure that the `DBClient` contract is being fulfilled correctly, because there is an additional layer of business logic that sits in-between the actual database client(Redis/Mongo) and the caller of the `DBClient`. I think the earlier comments in this issue point that out, by having situations that are not clearly defined and a divergence between implementations(Redis and Mongo). Even though we do have blackbox tests to exercise the system to ensure everything works are a feature level, it sometimes will not be able to test everything, because they are blackbox tests which are suppose to be built with no knowledge of internal behaviour. I am in favor of having unit and integration tests within this repo, as long as there is a good clean way to maintain and run the tests.

**Comment by brandonforster** [comment]
> If you can address the concerns from earlier in the thread wrt being able to easily unit test / smoke test when working in the persistence code, I'm happy to switch to the blackbox tests to as I go through dev/debug cycles.

The best thing I can offer for you today is that I've updated the blackbox documentation with instructions on how to run several test suites as individual Newman queries against local dev (as opposed to the old methods, which were either "test everything in Nexus against Newman or everything locally against Postman"

I've found that with some trivial shell scripts I can execute the blackbox tests very quickly and easily. I recognize that's not a fantastic answer, but it is a pragmatic one.

**Comment by andresrinivasan** [comment]
I can work with documentation and shell scripts (which hopefully are also documented and in the repo as well).

@AnthonyMBonafide I'm not following what you're proposing.

**Comment by AnthonyMBonafide** [comment]
My last comment was on response to @brandonforster 's comment. Basically, I like that the tests exist and some of his concerns can be addressed with a little refactor. That is just a suggestion, I didn't did enough to see how to refactor nor if it is even feasible let alone worth it. I apologize, I should have quoted his comment in my response to add some clarity to my statements.

**Comment by michaelestrin** [comment]
@brandonforster Let's discuss and plan a path forward when you return from vacation.

**Comment by tsconn23** [comment]
@andresrinivasan says we can close this.
